### PR TITLE
[Benchmark] Memory benchmark utils

### PR DESCRIPTION
**Pull request by patrickvonplaten**
## Description

Currently, we have a benchmarks file in examples `examples/benchmarks.py` which measures time and speed for inference and one `src/transformers/benchmark_utils.py` file which including the memory tracing from @thomwolf.

This PR generalizes the benchmarks much more:
- We have `benchmark_args_utils, benchmark_args, benchmark_tf_args` similar to `trainer_args`. Because there is such a big overlap between tf and torch, I decided to have a`benchmark_args_utils`
- We have a `benchmark_utils` including the previous code of `benchmark_utils` and a new abstract`Benchmark` class. `PyTorchBenchmark` and `TensorflowBenchmark` inherit from this class and implement the abstract methods `inference` and `train`. 
- Instead of being able to just measure our "default" models identified by their `model_name`, we can create a special config and measure the memory and time usage of the corresponding model before starting to train
- In addition to inference we can also measure train
- **IMPORTANT**: The line-by-line tracing implemented by @thomwolf is super useful to optimize one's code and see exactly which line produces most of the code, but it does not give the peek memory usage as one can see in `nvidia-smi` and which is relevant to decide how much RAM is needed for the model. I decided to not use the line-by-line tracing to plot the overall memory usage on GPU. It should rather only be used when one want to optimize his/her code instead of benchmarking the model. For Benchmarks the PyTorch method `torch.cuda.max_memory_reserved` is a better fit - it's much faster and gives the correct value.
- Also I added a `run_script` and a `plot_script` to the examples.
- @julien-c - I think it would be super nice to have such plots on our website somewhere. If people can see quickly how much more efficient model x is over y in for a given batch size and sequence length, I think it would be super useful.

## TODO:
- [x] Write tests
- [x] Make pretty (more type hints + smaller refactoring)

## Review

@thomwolf, @julien-c - It would be awesome if you could take a look here and check if you are ok with the general design choice. To get a better feeling how the Benchmark can be used, you can checkout this notebook: https://colab.research.google.com/drive/1GM3WPXmZ5wZHwhdPyf7RykJS3Y12iceJ?usp=sharing

## UPDATE

I think the PR is ready for merge now. The notebook: https://colab.research.google.com/drive/1GM3WPXmZ5wZHwhdPyf7RykJS3Y12iceJ?usp=sharing is updated and shows what environment and version infos are shown @thomwolf . Tests are included. The next PR should take care of pytorch jit tracing and TPU and TF @LysandreJik @julien-c 

## Future PR:
- [ ] Support Pytorch Tracing (discuss with Lysandre a bit before)
- [ ] Support TPU (discuss with Lysandre a bit before)
- [ ] TF version

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
This google colab shows nicely how the new benchmarking utils can be used: https://colab.research.google.com/drive/1GM3WPXmZ5wZHwhdPyf7RykJS3Y12iceJ?usp=sharing

**Comment by patrickvonplaten** [comment]
A plotter file is included in `examples/benchmarking` which can give the following results:

### Memory

memory_example

### Time

time_example

**Comment by sshleifer** [comment]
Can `csv_memory_filename_inference` be called `save_csv_path`?

File: src/transformers/benchmark/benchmark_utils.py [hunk 1]
```diff
@@ -307,6 +321,16 @@ def stop_memory_tracing(
                     frame=frame, cpu=Memory(cpu_mem_inc), gpu=Memory(gpu_mem_inc), cpu_gpu=Memory(cpu_gpu_mem_inc),
                 )
             )
+
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark_utils.py:324]: We add a current memory, which does not measure how much a line has added to the memory but how much memory is needed in total by this line. @thomwolf - I noticed that on GPU the highest `py3nvml` according to this code actually does not give the same results as `nvidia-smi`. I think the reason is that in order to measure the consumption per line we constantly reinitialize the memory tracing with `py3nvml.nvmlInit()` and `py3nvml.nvmlShutdown()` so that we don't trace added memory + currently hold memory. Also I think we would have to trace not only `transformers` code line by line but also all `PyTorch` code in order to trace the real memory peak usage. This is to slow though. Therefore I decided in order to measure the peak memory (which is usually the most important metric) to use PyTorch's `torch.cuda.max_memory_reserved` method which correspondns 1-to-1 to what is shown in `nvidia-smi`.

File: src/transformers/benchmark/benchmark_utils.py [hunk 2]
```diff
@@ -339,3 +375,177 @@ def bytes_to_human_readable(memory_amount: int) -> str:
...
+                self.print_results(train_result_time)
+                self.save_to_csv(train_result_time, self.args.csv_time_filename_train)
+
+            if not self.args.no_memory:
+                self.print_fn(f"======= TRAIN - MEMORY - RESULT =======")
+                self.print_results(train_result_memory)
+                self.save_to_csv(train_result_memory, self.args.csv_memory_filename_train)
+
+    def get_print_function(self, args):
+        if args.log_print:
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark_utils.py:470]: having this option allows to log all print outputs in a file. This can be very useful when applying line-by-line tracing.

File: src/transformers/hf_argparser.py [hunk 3]
```diff
@@ -65,6 +65,14 @@ def _add_dataclass_arguments(self, dtype: DataClassType):
                 if field.default is True:
                     field_name = f"--no-{field.name}"
                     kwargs["dest"] = field.name
+            elif hasattr(field.type, "__origin__") and issubclass(field.type.__origin__, List):
```
Comment (patrickvonplaten) [inline src/transformers/hf_argparser.py:68]: @julien-c - I need that for `list` inputs to the `HfArgParser`.

File: src/transformers/benchmark/benchmark_tf_args.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+if is_tf_available():
+    import tensorflow as tf
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class TensorflowBenchmarkArguments(BenchmarkArguments):
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark_tf_args.py:32]: Not really implemented yet

File: src/transformers/benchmark/benchmark.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        model.to(self.args.device)
+        model.train()
+
+        input_ids = torch.randint(
+            model.config.vocab_size, (batch_size, sequence_length), dtype=torch.long, device=self.args.device
+        )
+
+        def compute_loss_and_backprob():
+            # TODO: Not all models call labels argument labels => this hack using the function signature should be corrected once all models have a common name for labels
+            function_argument_names = inspect.getfullargspec(model.forward).args
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark.py:49]: @julien-c - This is currently done because we have different names for `labels` (I think `Trainer` has the same issue).

File: src/transformers/benchmark/benchmark.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            runtimes = timeit.repeat(
+                lambda: compute_loss_and_backprob(),
+                repeat=self.args.repeat,
+                number=3,
+            )
+            return min(runtimes) / 3.0
+
+    def inference(self, model_name, batch_size, sequence_length, trace_memory=False):
+        config = self.config_dict[model_name]
+        model = MODEL_MAPPING[config.__class__](config)
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark.py:91]: We only load the model into memory here as it can consume a significant chunk of the memory.

File: src/transformers/benchmark/benchmark_utils.py [hunk 7]
```diff
@@ -339,3 +375,177 @@ def bytes_to_human_readable(memory_amount: int) -> str:
...
+
+    args: BenchmarkArguments
+    configs: PretrainedConfig
+    print_fn: Callable[[str], None]
+
+    def __init__(self, args: BenchmarkArguments = None, configs: PretrainedConfig = None):
+        self.args = args
+        self.print_fn = self.get_print_function(args)
+
+        if configs is None:
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark_utils.py:394]: The idea is here the following: 
We want to allow both measuring time / memory of our "standard" models, but also of specific models given just their config. If the config is None => we use the standard models according to the `model_names` else we init the model class with a specific config.

File: src/transformers/benchmark/benchmark_utils.py [hunk 8]
```diff
@@ -321,16 +345,28 @@ def stop_memory_tracing(
...
-        return MemorySummary(sequential=memory_diff_trace, cumulative=cumulative_memory, total=total_memory)
+
+        return MemorySummary(
+            sequential=memory_diff_trace, cumulative=cumulative_memory, current=memory_curr_trace, total=total_memory,
+        )
 
     return None
 
 
+def bytes_to_mega_bytes(memory_amount: int) -> int:
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark_utils.py:362]: IMO, better to have a readable int here

**Comment by patrickvonplaten** [comment]
> Can `csv_memory_filename_inference` be called `save_csv_path`?

Yeah about the exact naming I think we can decide a bit later. Wanna see first if you guys are ok with the general design. But we have up to four csv files that are saved - so `save_csv_path` is too general.

File: src/transformers/benchmark/benchmark.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            if self.args.trace_memory_line_by_line or self.args.n_gpu == 0:
+                summary = stop_memory_tracing(trace)
+                memory = summary.total
+            else:
+                memory = Memory(torch.cuda.max_memory_reserved())
+
+            return memory
+        else:
+            # as written in https://docs.python.org/2/library/timeit.html#timeit.Timer.repeat, min should be taken rather than the average
+            runtimes = timeit.repeat(
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark.py:82]: https://docs.python.org/2/library/timeit.html#timeit.Timer.repeat says that one should rather take the `min()` than the `average()` @thomwolf .

Also maybe we might want to increase `number` from 3 a bit (maybe 10?) since for smaller sequence lengths there seems to be very little difference in running time.

**Review by thomwolf — APPROVED** [review]
LGTM!

As discussed offline, I think that for good reproducibility, the benchmark script (or even the library benchmark_utils? I let you choose) should also save a dict/JSON with all the parameters used for the benchmark (eg training versus no training, on GPU versus on CPU) as well as PyTorch and transformers versions, maybe all the PyPi versions of the installed packages actually, and the date and the type of GPU and hardware if you can get it. This can be quite important to understand/investigate differences or regression in performances that could happen in the future.

We have a `transformer-cli env` utility that you could want to leverage for that.

File: examples/benchmarking/plot_csv_file.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        metadata={
+            "help": "Filename under which the plot will be saved. Only relevant if `save_fig` is True."
+        }
+    )
+
+
+class Plot:
+    def __init__(self, args):
+        self.args = args
+        self.result_dict = {}
```
Comment (sshleifer) [inline examples/benchmarking/plot_csv_file.py:48]: if you use `defaultdict(dict)` you can save the `if` on L54

File: examples/benchmarking/plot_csv_file.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+        with open(self.args.csv_file, newline="") as csv_file:
+            reader = csv.DictReader(csv_file)
+            for row in reader:
+                model_name = row["model"]
+                if model_name not in self.result_dict:
+                    self.result_dict[model_name] = {}
```
Comment (sshleifer) [inline examples/benchmarking/plot_csv_file.py:55]: simpler IMO
```python
empty_record = lambda: dict(bsz=[], seq_len=[], result={})
self.result_dict = defaultdict(empty_record)
```

File: src/transformers/benchmark/benchmark.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+                torch.cuda.reset_peak_memory_stats()
+
+            # calculate loss and do backpropagation
+            compute_loss_and_backprob()
+
+            if self.args.trace_memory_line_by_line or self.args.n_gpu == 0:
+                summary = stop_memory_tracing(trace)
+                memory = summary.total
+            else:
+                memory = Memory(torch.cuda.max_memory_reserved())
```
Comment (sshleifer) [inline src/transformers/benchmark/benchmark.py:77]: document that this only supports `torch v1.5+`?

File: src/transformers/benchmark/benchmark_utils.py [hunk 13]
```diff
@@ -62,14 +69,14 @@ class UsedMemoryState(NamedTuple):
 
 class Memory(NamedTuple):
```
Comment (sshleifer) [inline src/transformers/benchmark/benchmark_utils.py:70]: -> `MemRecording` /`MemData` This is a very abstract name

File: src/transformers/benchmark/benchmark_utils.py (hunk 2 above)
Comment (sshleifer) [reply src/transformers/benchmark/benchmark_utils.py:469]: @property since you already have `self.args`?

File: src/transformers/benchmark/benchmark_utils.py [hunk 14]
```diff
@@ -339,3 +374,177 @@ def bytes_to_human_readable(memory_amount: int) -> str:
...
+                    if isinstance(result, float):
+                        self.print_fn(
+                            f"\t\t{model_name}/{batch_size}/{sequence_length}: " f"{(round(1000 * result) / 1000)}s"
+                        )
+                    else:
+                        self.print_fn(f"\t\t{model_name}/{batch_size}/{sequence_length}: " f"{result} MB")
+
+    def print_memory_trace_statistics(self, summary: MemorySummary):
+        self.print_fn(
+            "\nLines by line memory consumption:\n"
```
Comment (sshleifer) [inline src/transformers/benchmark/benchmark_utils.py:500]: typo -> line by line

File: src/transformers/benchmark/benchmark_utils.py [hunk 15]
```diff
@@ -339,3 +374,177 @@ def bytes_to_human_readable(memory_amount: int) -> str:
...
+            + "\n".join(
+                f"{state.frame.filename}:{state.frame.line_number}: mem {state.cpu_gpu}: {state.frame.line_text}"
+                for state in summary.sequential
+            )
+        )
+        self.print_fn(
+            "\nLines with top memory consumption:\n"
+            + "\n".join(
+                f"=> {state.frame.filename}:{state.frame.line_number}: mem {state.cpu_gpu}: {state.frame.line_text}"
+                for state in summary.cumulative[:6]
```
Comment (sshleifer) [inline src/transformers/benchmark/benchmark_utils.py:510]: kwarg 6?

**Review by sshleifer — APPROVED** [review]
This is really nice. I left some nitpicks that don't matter much!

File: src/transformers/benchmark/benchmark.py (hunk 5 above)
Comment (LysandreJik) [reply src/transformers/benchmark/benchmark.py:49]: Yes, I would be in favor of that!

File: src/transformers/benchmark/benchmark_args_utils.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+def list_field(default=None, metadata=None):
+    return field(default_factory=lambda: default, metadata=metadata)
+
+
+@dataclass
+class BenchmarkArguments:
+    """
+    BenchMarkArguments are arguments we use in our benchmark scripts
+    **which relate to the training loop itself**.
```
Comment (julien-c) [inline src/transformers/benchmark/benchmark_args_utils.py:32]:
```suggestion
```

delete this line

File: examples/benchmarking/run_benchmarks.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Benchmarking the library on inference and training """
+
+# If checking the tensors placement
+# tf.debugging.set_log_device_placement(True)
```
Comment (julien-c) [inline examples/benchmarking/run_benchmarks.py:19]: maybe remove this and add in doc instead

File: src/transformers/benchmark/benchmark_args_utils.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+                "transfo-xl-wt103",
+                "openai-gpt",
+                "distilbert-base-uncased",
+                "distilgpt2",
+                "roberta-base",
+                "ctrl",
+                "t5-base",
+                "facebook/bart-large",
+                "google/reformer-enwik8",
+            ]
```
[the remark this answers is not in the document]
Comment (julien-c) [reply src/transformers/benchmark/benchmark_args_utils.py:106]: You can also do a __post_init__ if necessary

File: src/transformers/hf_argparser.py (hunk 3 above)
Comment (julien-c) [reply src/transformers/hf_argparser.py:68]: LGTM, but can you also add the corresponding tests to `test_hf_argparser.py`?

File: examples/benchmarking/plot_csv_file.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        metadata={
+            "help": "Whether the plot should be saved or displayed directly with `plt.show()`. Defaults to directly displaying the plot."
+        },
+    )
+    figure_file: str = field(
+        default="plot.png",
+        metadata={
+            "help": "Filename under which the plot will be saved. Only relevant if `save_fig` is True."
+        }
+    )
```
Comment (julien-c) [inline examples/benchmarking/plot_csv_file.py:42]: maybe group those two in one option, Optional[str]

File: src/transformers/benchmark/benchmark.py [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+
+from .benchmark_utils import Benchmarks, Memory, start_memory_tracing, stop_memory_tracing
+
+
+if is_torch_available():
+    import torch
+    from .benchmark_args import PyTorchBenchmarkArguments
+
+
+class PyTorchBenchmarks(Benchmarks):
```
Comment (julien-c) [inline src/transformers/benchmark/benchmark.py:32]:
```suggestion
class PyTorchBenchmark(Benchmark):
```

nit, but can those classes be singular instead of plural?

File: src/transformers/benchmark/benchmark.py (hunk 5 above)
Comment (julien-c) [reply src/transformers/benchmark/benchmark.py:49]: would be great

**Review by julien-c — APPROVED** [review]
Could be nice to integrate this in CI too (regression testing) at some point, or to automatically feed a big spreadsheet like @LysandreJik's https://docs.google.com/spreadsheets/d/1sryqufw2D0XlUH4sq3e9Wnxu5EAQkaohzrJbd5HdQ_w/edit#gid=0 over time

File: src/transformers/benchmark/benchmark_args_utils.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/benchmark/benchmark_args_utils.py:106]: True it shouldn't be here - I will move it out to the experiments folder

File: src/transformers/hf_argparser.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/hf_argparser.py:72]: > Should `List` also allow `Tuple`?

I think it's hard to parse a list like object from the command line like `--tuple_arg 1 2 3`  with `argparse.parse` that will not come out as a list but as a tuple. Also I don't really see the need to be able to parse tuples if one can parse lists in data classes. It would be easy to allow tuples in dataclasses to be parsed via the command line as a list, but then why would one want to parse something into a tuple and not into  a list... => so I decided to just stick to list for now

File: src/transformers/benchmark/benchmark.py (hunk 12 above)
Comment (patrickvonplaten) [reply src/transformers/benchmark/benchmark.py:82]: It since 1.4+ => but will document this

File: src/transformers/benchmark/benchmark_args.py [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+def is_tpu_available():
+    return _has_tpu
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class PyTorchBenchmarkArguments(BenchmarkArguments):
+    no_cuda: bool = field(default=False, metadata={"help": "Weather to run on available cuda devices"})
```
Comment (LysandreJik) [inline src/transformers/benchmark/benchmark_args.py:46]:
```suggestion
    no_cuda: bool = field(default=False, metadata={"help": "Whether to run on available cuda devices"})
```

File: src/transformers/benchmark/benchmark_args_utils.py [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+
+    Using `HfArgumentParser` we can turn this class
+    into argparse arguments to be able to specify them on
+    the command line.
+    """
+
+    models: List[str] = list_field(
+        default=[],
+        metadata={
+            "help": "Model checkpoints to be provided to the AutoModel classes. Leave blank to benchmark the base version of all available model "
```
Comment (LysandreJik) [inline src/transformers/benchmark/benchmark_args_utils.py:42]:
```suggestion
            "help": "Model checkpoints to be provided to the AutoModel classes. Leave blank to benchmark the base version of all available models"
```

File: src/transformers/benchmark/benchmark.py [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+                    memory = Memory(torch.cuda.max_memory_reserved())
+                else:
+                    logger.info("Please consider updating PyTorch to version 1.4 to get more accuracy on GPU memory usage")
+                    memory = Memory(torch.cuda.max_memory_cached())
+
+            return memory
+        else:
+            # as written in https://docs.python.org/2/library/timeit.html#timeit.Timer.repeat, min should be taken rather than the average
+            runtimes = timeit.repeat(lambda: model(input_ids), repeat=self.args.repeat, number=10,)
+            return min(runtimes) / 10.0
```
Comment (LysandreJik) [inline src/transformers/benchmark/benchmark.py:133]: Might have been there before your refactor, but I believe it's usually better to time the average rather than the minimum. The average takes into account hardware stability, and is less likely to bias the results. 

I would advocate for the more stable approach, knowing how the model will usually perform on a given hardware, rather than the absolute best the model will perform, sampled from 10 runs. Open to criticism though, I had opted for minimum as well in the first version ([original benchmark script before TensorFlow team and François Chollet's comments](https://gist.github.com/LysandreJik/c174a4fdc43004f5a2389f81d11444cd/5362a9236ed1d1739df7caa9e0d83a65ee07f2d0)), before switching back to mean.

**Review by LysandreJik — APPROVED** [review]
This is really cool, I'm looking forward to using it.

It seems you've implemented TPU support as well, did you get a chance to try it?

File: src/transformers/benchmark/benchmark.py (hunk 23 above)
Comment (LysandreJik) [reply src/transformers/benchmark/benchmark.py:133]: On a second note, you seem to be running inference on 10 runs right after initializing the model. This isn't problematic in most cases, but would be for cases where the model must be traced before being run (see TorchScript, TPU).

On TPU for example, it's not surprising to have the first iteration run for 45 seconds, and the next iterations at 0.3s/it. The first iteration would largely skew the results, while this first iteration would be meaningless in a production setup where the model would only be traced during initialization.

Comment (patrickvonplaten) [reply src/transformers/benchmark/benchmark.py:137]: In https://docs.python.org/2/library/timeit.html#timeit.Timer.repeat , it is written that `min()` should be used instead of average. 
I think the way it is implemented now is that it is run `self.args.repeat` times 10 times in a row and then ouf of those `self.args.repeat` times the min value is taken and average of the 10 times. 

I'm really not sure, whether it's better to time average or `min`. I had average in the beginning as well, but after reading  https://docs.python.org/2/library/timeit.html#timeit.Timer.repeat I switched to min. Since it is still averaged over 10 runs I guess `min()` is fine here as well, but I think you have more experience, so I let you decide :-) 

Regarding the second note, I think we should have a different timing mechanism for TPU. Maybe a simple `if` in the code is enough?

**Comment by patrickvonplaten** [comment]
> This is really cool, I'm looking forward to using it.
> 
> It seems you've implemented TPU support as well, did you get a chance to try it?

I didn't play around with TPU very much yet. I'd like to merge this PR which works for GPU / CPU for PyTorch now and then maybe clean things for TPU in a second PR?

File: src/transformers/benchmark/benchmark.py (hunk 23 above)
Comment (LysandreJik) [reply src/transformers/benchmark/benchmark.py:137]: Right, I guess that's okay. After reading a bit more on the subject, it seems that the choice really depends on the distribution so I don't have any particular insights. The average 10 times before minimum seems robust enough to me.

> Regarding the second note, I think we should have a different timing mechanism for TPU. Maybe a simple if in the code is enough?

Right, this is especially noticeable for TPUs, but actually really impacts the TorchScript performance as well. From what I'm seeing it doesn't seem you're actually leveraging torchscript/tracing the model anywhere, so it's probably not necessary yet. 

Maybe a `requires_tracing` flag for inference would be useful, where it does a few inferences at first before actually evaluating time spent, to make sure everything is correctly traced beforehand. Would be welcome in a second PR!

**Review by thomwolf — APPROVED** [review]
LGTM
